tensorly/decomposition/_nn_cp.py

```diff
@@ -0,0 +1,413 @@
+
+    elif init == 'svd':
+        try:
+            svd_fun = tl.SVD_FUNS[svd]
+        except KeyError:
+            message = 'Got svd={}. However, for the current backend ({}), the possible choices are {}'.format(
+                    svd, tl.get_backend(), tl.SVD_FUNS)
+            raise ValueError(message)
+
+        factors = []
+        for mode in range(tl.ndim(tensor)):
+            U, S, V = svd_fun(unfold(tensor, mode), n_eigenvecs=rank)
+
+            # NNDSVD initialization
+            W = tl.zeros_like(U)
+            H = tl.zeros_like(V)
+
+            # The leading singular triplet is non-negative
+            # so it can be used as is for initialization.
+            W = tl.index_update(W, tl.index[:, 0], tl.sqrt(S[0]) * tl.abs(U[:, 0]))
+            H = tl.index_update(H, tl.index[0, :], tl.sqrt(S[0]) * tl.abs(V[0, :]))
+
+            for j in range(1, tl.shape(U)[1]):
+                x, y = U[:, j], V[j, :]
+
+                # extract positive and negative parts of column vectors
+                x_p, y_p = tl.clip(x, a_min=0.0), tl.clip(y, a_min=0.0)
+                x_n, y_n = tl.abs(tl.clip(x, a_max=0.0)), tl.abs(tl.clip(y, a_max=0.0))
+
+                # and their norms
+                x_p_nrm, y_p_nrm = tl.norm(x_p), tl.norm(y_p)
+                x_n_nrm, y_n_nrm = tl.norm(x_n), tl.norm(y_n)
+
+                m_p, m_n = x_p_nrm * y_p_nrm, x_n_nrm * y_n_nrm
+
+                # choose update
+                if m_p > m_n:
+                    u = x_p / x_p_nrm
+                    v = y_p / y_p_nrm
+                    sigma = m_p
+                else:
+                    u = x_n / x_n_nrm
+                    v = y_n / y_n_nrm
+                    sigma = m_n
+
+                lbd = tl.sqrt(S[j] * sigma)
+                W = tl.index_update(W, tl.index[:, j], lbd * u)
+                H = tl.index_update(H, tl.index[j, :], lbd * v)
+
+            # After this point we no longer need H
+            idx = W < np.finfo(np.float32).eps
+
+            if nntype == "nndsvd":
+                W = tl.index_update(W, tl.index[idx], 0.0)
+            elif nntype == "nndsvda":
+                avg = tl.mean(tensor)
+                W = tl.index_update(W, tl.index[idx], avg)
+            else:
+                raise ValueError(
+                    'Invalid nntype parameter: got %r instead of one of %r' %
+                    (init, ('nndsvd', 'nndsvda')))
+
+            U = W
+
+            if tensor.shape[mode] < rank:
+                # TODO: this is a hack but it seems to do the job for now
+                random_part = tl.tensor(rng.random_sample((U.shape[0], rank - tl.shape(tensor)[mode])), **tl.context(tensor))
+                U = tl.concatenate([U, random_part], axis=1)
+
+            factors.append(U[:, :rank])
+
+        kt = CPTensor((None, factors))
+
+    elif isinstance(init, (tuple, list, CPTensor)):
+        # TODO: Test this
+        try:
+            kt = CPTensor(init)
+        except ValueError:
```

Comment:
Should this exception be within CPTensor?

Reply:
Possibly but in this case we also want to alert the user about what he did wrong (e.g. they should get an error message explaining what is wrong -- not a CP init-- and hinting at how to fix it. If the error is raised in CPTensor it might be harder for them to understand and fix the issue?